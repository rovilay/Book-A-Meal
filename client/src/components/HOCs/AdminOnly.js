import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import jwt from 'jsonwebtoken';

import isExpired from '../../helpers/isExpired';
import {
  getFromLocalStorage,
  deleteInLocalStorage
} from '../../helpers/localstorage';
import { setDefaultNav, setNav } from '../../actions/navLinksActions';

/**
 * @export {function} HOC function that returns
  a component if user is admin/caterer
 * @param  {Component} Comp Component
 * @return {Component} only if user is a caterer/admin
 */
export default function (Comp) {
  class AdminOnly extends Component {
    constructor(props) {
      super(props);
      this.state = {
        token: '',
        admin: undefined
      };
    }

    /* istanbul ignore next */
    componentWillMount() {
      const { history } = this.props;
      const token = getFromLocalStorage('jwt');
      if (token) {
        const decodedToken = jwt.decode(token);
        if (!decodedToken) {
          deleteInLocalStorage('jwt');
          return history.push('/login');
        }

        const {
          exp,
          admin
        } = decodedToken;

        if (!exp || !admin) {
          deleteInLocalStorage('jwt');
          this.props.setDefaultNav();
          return history.push('/login');
        }

        if (!admin) {
          this.props.setDefaultNav();
          return history.push('/');
        }

        if (exp && isExpired(exp)) {
          this.props.setDefaultNav();
          return history.push('/login');
        }

        this.setState({ token, admin });
      }

      if (!token) {
        this.props.setDefaultNav();
        return history.push('/');
      }
    }

    render() {
      const { token, admin } = this.state;
      return (
        <div>
          {
            (admin) && <Comp {...this.props} token={token} />
          }
        </div>
      );
    }
  }

  AdminOnly.propTypes = {
    history: PropTypes.object.isRequired,
    setDefaultNav: PropTypes.func.isRequired,
    setNav: PropTypes.func.isRequired,
  };

  const mapDispatchToProps = dispatch => bindActionCreators(
    {
      setDefaultNav,
      setNav,
    },
    dispatch
  );

  return connect('', mapDispatchToProps)(withRouter(AdminOnly));
}
