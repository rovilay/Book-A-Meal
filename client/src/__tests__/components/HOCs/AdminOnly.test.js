import React, { Component } from 'react';
import { shallow } from 'enzyme';
import mockStore from '../../__mockData__/mockStore';
import { menuMeals } from '../../__mockData__/menuMock';
import toJson from 'enzyme-to-json';
import AdminOnly from '../../../components/HOCs/AdminOnly';
import MealPage from '../../../components/adminpages/mealPage/Index'


describe('AdminOnly higher order component test', () => {
  let WrapperWithHOC;

  // create mock components
  beforeEach(() => {
    class MockComponent extends Component {
      render() {
        return (<div>Component</div>)
      }
    };

    WrapperWithHOC = AdminOnly(MockComponent);
  })

  const setup = () => {
    const props = {
      history: [],
      setDefaultNav: () => jest.fn(),
      setNav: () => jest.fn()
    };

    const store = mockStore({
      menus: {
        todayMenu: menuMeals
      }
    })

    return shallow(<WrapperWithHOC  {...props} store={store} />)
  };


  it('should return component only if admin', (done) => {
    localStorage.setItem('jwt', adminToken);
    const wrapper = setup()

    expect(toJson(wrapper)).toMatchSnapshot();

    done();
  });


  it('should return component only if admin', (done) => {
    const wrapper = setup()

    expect(toJson(wrapper)).toMatchSnapshot();

    done();
  });
});
