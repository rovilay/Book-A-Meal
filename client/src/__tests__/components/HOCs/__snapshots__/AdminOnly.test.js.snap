// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminOnly higher order component test should return component only if admin 1`] = `
<withRouter(AdminOnly)
  history={Array []}
  setDefaultNav={[Function]}
  setNav={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;

exports[`AdminOnly higher order component test should return component only if admin 2`] = `
<withRouter(AdminOnly)
  history={Array []}
  setDefaultNav={[Function]}
  setNav={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
